### PR TITLE
XGBoost Classifier

### PR DESCRIPTION
**Pull request by Moddy2024**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by Moddy2024** [comment]
@cclauss  Can you please review this I hope this will be merged as all the tests are passing.

**Comment by cclauss** [comment]
Your only function, `main()` takes no inputs and returns no outputs and it plots.  This is not an algorithm by this repo's definition.

https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md#what-is-an-algorithm
#### What is an Algorithm?

An Algorithm is one or more functions (or classes) that:
* take one or more inputs,
* perform some internal calculations or data manipulations,
* return one or more outputs,
* have minimal side effects (Ex. `print()`, `plot()`, `read()`, `write()`).

**Comment by Moddy2024** [comment]
> Your only function, `main()` takes no inputs and returns no outputs and it plots. This is not an algorithm by this repo's definition.
> 
> https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md#what-is-an-algorithm
> 
> #### What is an Algorithm?
> An Algorithm is one or more functions (or classes) that:
> 
> * take one or more inputs,
> * perform some internal calculations or data manipulations,
> * return one or more outputs,
> * have minimal side effects (Ex. `print()`, `plot()`, `read()`, `write()`).

But there are same kind of algorithms which already exists in your repo

**Comment by cclauss** [comment]
Why not open a pull request here?  https://github.com/dmlc/xgboost/tree/master/demo  They probably want how-to-use examples.  We can to focus on algorithms.

**Comment by Moddy2024** [comment]
> Why not open a pull request here? https://github.com/dmlc/xgboost/tree/master/demo They probably want how-to-use examples. We want to focus on algorithms.

This [algorithm](https://github.com/TheAlgorithms/Python/blob/master/machine_learning/knn_sklearn.py) does not have any functions at all but it is there in this repository.

**Comment by cclauss** [comment]
Yes.  That was merged 1,196 days ago and we have raised our standards for new submissions significantly since then.  There are lots of repos that are easier to contribute to.

**Comment by Moddy2024** [comment]
> Your only function, `main()` takes no inputs and returns no outputs and it plots. This is not an algorithm by this repo's definition.
> 
> https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md#what-is-an-algorithm
> 
> #### What is an Algorithm?
> An Algorithm is one or more functions (or classes) that:
> 
> * take one or more inputs,
> * perform some internal calculations or data manipulations,
> * return one or more outputs,
> * have minimal side effects (Ex. `print()`, `plot()`, `read()`, `write()`).

I have edited the script according to the deifinition os this repo now

**Comment by Moddy2024** [comment]
@cclauss  Hello it's been 5 day since I made the changes can you review this PR whenever you get time.
Thank You

**Comment by cclauss** [comment]
[These are type hints.](https://github.com/TheAlgorithms/Python/blob/master/maths/add.py#L6)   If your function parameters do not have them then your pull request will not be reviewed and it will not be merged.

[These are doctests.](https://github.com/TheAlgorithms/Python/blob/master/maths/add.py#L7-L12)  If your functions do not have them then your pull request will not be reviewed and it will not be merged.

Please read [`CONTRIBUTING.md`](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md) for details.

File: machine_learning/xgboostclassifier.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+from xgboost import XGBClassifier
+
+
+def data_handling(data: dict) -> tuple:
+    # Split dataset into train and test data
+    # data is features
+    """
+    >>> data_handling(({'data':'[5.1, 3.5, 1.4, 0.2]','target':([0])}))
+    ('[5.1, 3.5, 1.4, 0.2]', [0])
+    >>> data_handling({'data':'[4.9, 3. , 1.4, 0.2],[4.7, 3.2, 1.3, 0.2]','target':([0, 0])})
```
Comment (cclauss) [inline machine_learning/xgboostclassifier.py:16]: machine_learning/xgboostclassifier.py:16:89: E501 line too long (93 > 88 characters)
```suggestion
    >>> data_handling(
    ...     {'data':'[4.9, 3. , 1.4, 0.2],[4.7, 3.2, 1.3, 0.2]','target':([0, 0])}
    ... )
```

File: machine_learning/xgboostclassifier.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    ('[4.9, 3. , 1.4, 0.2],[4.7, 3.2, 1.3, 0.2],[4.6, 3.1, 1.5, 0.2]', [0, 0, 0])
+    """
+    x = (data["data"], data["target"])
+    return x
```
Comment (cclauss) [inline machine_learning/xgboostclassifier.py:20]:
```suggestion
    return (data["data"], data["target"])
```

File: machine_learning/xgboostclassifier.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    names = iris["target_names"]
+
+    # XGBoost Classifier
+    xgb = xgboost(x_train, y_train)
+
+    # Display Confusion Matrix of Classifier
+    # with both train and test sets
+    plot_confusion_matrix(
+        xgb,
```
Comment (cclauss) [inline machine_learning/xgboostclassifier.py:58]:
```suggestion
    # Create an XGBoost Classifier from the training data
    xgboost_classifier = xgboost(x_train, y_train)

    # Display the confusion matrix of the classifier with both the training and test sets
    plot_confusion_matrix(
        xgboost_classifier,
```

File: machine_learning/xgboostclassifier.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    plt.title("Normalized Confusion Matrix - IRIS Dataset")
+    plt.show()
+
+
+if __name__ == "__main__":
+    import doctest
+
+    doctest.testmod(name="main", verbose=True)
+    doctest.testmod(name="xgboost", verbose=True)
+    doctest.testmod(name="data_handling", verbose=True)
```
Comment (cclauss) [inline machine_learning/xgboostclassifier.py:74]:
```suggestion
    doctest.testmod(verbose=True)
```

File: machine_learning/xgboostclassifier.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    classifier.fit(features, target)
+    return classifier
+
+
+def main() -> None:
+
+    """
+    >>> main()
+
+    The Url for the algorithm
```
Comment (cclauss) [inline machine_learning/xgboostclassifier.py:38]:
```suggestion
    Url for the algorithm:
```

**Comment by cclauss** [comment]
OK.  I have added better names and wrapped long lines to placate pre-commit.

The only test that fails is one that I do not know how to solve.

Please click the `Details` button on the line with the ❌ below to see the error.  You can run these tests on your own machine with the command: `python3 -m doctest -v machine_learning/xgboostclassifier.py`

**Comment by Moddy2024** [comment]
> OK. I have added better names and wrapped long lines to placate pre-commit.
> 
> The only test that fails is one that I do not know how to solve.
> 
> Please click the `Details` button on the line with the ❌ below to see [the error.](https://github.com/TheAlgorithms/Python/actions/runs/3287466564/jobs/5416668576#step:6:587) You can run these tests on your own machine with the command: `python3 -m doctest -v machine_learning/xgboostclassifier.py`
> 
> UNEXPECTED EXCEPTION: ValueError('Invalid classes inferred from unique values of `y`. Expected: [0 1], got [1 2]')

Thank you very much. That's so kind of you.

**Comment by Moddy2024** [comment]
@cclauss  Thank you with your help I removed the errors.

**Review by cclauss — APPROVED** [review]
Thanks massively for your persistence!!!

**Comment by Moddy2024** [comment]
> Thanks massively for your persistence!!!

Thank you for your guidance and help.

**Pull request merged**